plugins/bulk-renamer/BulkRenamer.vala

```diff
@@ -0,0 +1,568 @@
+/*
+ * Copyright (C) 2010-2017  Vartan Belavejian
+ * Copyright (C) 2019-2020     Jeremy Wootten
+ *
+    This program is free software: you can redistribute it and/or modify
+    it under the terms of the GNU General Public License as published by
+    the Free Software Foundation, either version 3 of the License, or
+    (at your option) any later version.
+
+    This program is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License
+    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ *  Authors:
+ *  Vartan Belavejian <https://github.com/VartanBelavejian>
+ *  Jeremy Wootten <jeremywootten@gmail.com>
+ *
+*/
+
+public class Files.Renamer : Gtk.Grid {
+    private Gee.HashMap<string, GLib.File> file_map;
+    private Gee.HashMap<string, FileInfo> file_info_map;
+    private Gee.ArrayList<Modifier> modifier_chain;
+
+    private Gtk.Grid modifier_grid;
+    private Gtk.ListBox modifier_listbox;
+
+    private Gtk.TreeView old_file_names;
+    private Gtk.TreeView new_file_names;
+    private Gtk.ListStore old_list;
+    private Gtk.ListStore new_list;
+    private Icon invalid_icon;
+
+    private Gtk.Entry base_name_entry;
+    private Gtk.ComboBoxText base_name_combo;
+    private Gtk.Switch sort_type_switch;
+    private Gtk.ComboBoxText sort_by_combo;
+
+    private Mutex info_map_mutex;
+
+    private int number_of_files = 0;
+
+    public bool can_rename { get; set; default = false; }
+
+    public string directory { get; private set; default = ""; }
+
+    public Renamer (GLib.File[]? files = null) {
+        if (files != null) {
+            add_files (files);
+        }
+    }
+
+    construct {
+        info_map_mutex = Mutex ();
+        invalid_icon = new ThemedIcon.with_default_fallbacks ("dialog-warning");
+        can_rename = false;
+        orientation = Gtk.Orientation.VERTICAL;
+        directory = "";
+
+        file_map = new Gee.HashMap<string, GLib.File> ();
+        file_info_map = new Gee.HashMap<string, FileInfo> ();
+        modifier_chain = new Gee.ArrayList<Modifier> ();
+
+        var base_name_label = new Granite.HeaderLabel (_("Base"));
+        base_name_label.get_style_context ().add_class (Granite.STYLE_CLASS_H2_LABEL);
+
+        base_name_combo = new Gtk.ComboBoxText () {
+            valign = Gtk.Align.CENTER
+        };
+        base_name_combo.insert (RenameBase.ORIGINAL, "ORIGINAL", RenameBase.ORIGINAL.to_string ());
+        base_name_combo.insert (RenameBase.CUSTOM, "CUSTOM", RenameBase.CUSTOM.to_string ());
+
+        base_name_entry = new Gtk.Entry () {
+            placeholder_text = _("Enter naming scheme"),
```

Comment:
Placeholder text should be an example of valid text for this entry. Descriptions of entries should be labels outside the entry. I'm not sure what good example text here would be, so we should probably remove this